Kernel/Graphics/GraphicsManagement.cpp

```diff
@@ -120,40 +120,30 @@ static inline bool is_display_controller_pci_device(PCI::DeviceIdentifier const&
     return device_identifier.class_code().value() == 0x3;
 }
 
-UNMAP_AFTER_INIT bool GraphicsManagement::determine_and_initialize_graphics_device(PCI::DeviceIdentifier const& device_identifier)
+struct PCIGraphicsDriverInitializer {
+    ErrorOr<bool> (*probe)(PCI::DeviceIdentifier const&) = nullptr;
+    ErrorOr<NonnullLockRefPtr<GenericGraphicsAdapter>> (*create)(PCI::DeviceIdentifier const&) = nullptr;
+};
+
+static constexpr PCIGraphicsDriverInitializer s_initializers[] = {
+    { IntelNativeGraphicsAdapter::probe, IntelNativeGraphicsAdapter::create },
+    { BochsGraphicsAdapter::probe, BochsGraphicsAdapter::create },
+    { VirtIOGraphicsAdapter::probe, VirtIOGraphicsAdapter::create },
+    { VMWareGraphicsAdapter::probe, VMWareGraphicsAdapter::create },
+};
+
+UNMAP_AFTER_INIT ErrorOr<void> GraphicsManagement::determine_and_initialize_graphics_device(PCI::DeviceIdentifier const& device_identifier)
 {
     VERIFY(is_vga_compatible_pci_device(device_identifier) || is_display_controller_pci_device(device_identifier));
-    LockRefPtr<GenericGraphicsAdapter> adapter;
-
-    if (!adapter) {
-        switch (device_identifier.hardware_id().vendor_id) {
-        case PCI::VendorID::QEMUOld:
-            if (device_identifier.hardware_id().device_id == 0x1111)
-                adapter = BochsGraphicsAdapter::initialize(device_identifier);
-            break;
-        case PCI::VendorID::VirtualBox:
-            if (device_identifier.hardware_id().device_id == 0xbeef)
-                adapter = BochsGraphicsAdapter::initialize(device_identifier);
-            break;
-        case PCI::VendorID::Intel:
-            adapter = IntelNativeGraphicsAdapter::initialize(device_identifier);
-            break;
-        case PCI::VendorID::VirtIO:
-            dmesgln("Graphics: Using VirtIO console");
-            adapter = VirtIOGraphicsAdapter::initialize(device_identifier);
-            break;
-        case PCI::VendorID::VMWare:
-            adapter = VMWareGraphicsAdapter::try_initialize(device_identifier);
-            break;
-        default:
-            break;
+    for (auto& initializer : s_initializers) {
+        auto initializer_probe_found_driver_match = TRY(initializer.probe(device_identifier));
```

Comment:
If for some reason probing for one of the graphics adapters failed and returned an error (not the case for the adapters right now, I know), can we `dmesgln` the error and continue with the next initializer?

Reply:
Done. :)